node-red-contrib-botbuilder/msbot/actions.js

```diff
@@ -1,32 +1,140 @@
-const { BotFrameworkAdapter, TurnContext } = require("botbuilder");
+const { BotFrameworkAdapter, TurnContext, MemoryStorage, ConversationState, SkillHttpClient, SkillHandler, ChannelServiceRoutes, ActivityTypes } = require("botbuilder");
+const { SimpleCredentialProvider, AuthenticationConfiguration } = require('botframework-connector');
 const { VBMBot } = require("./botVBM");
 const getMessage = require("./messages.js");
-
-
+const { SkillsConfiguration } = require('./skillsConfiguration');
+const { SkillConversationIdFactory } = require('./skillConversationIdFactory');
+const { AllowedCallersClaimsValidator } = require('./allowedCallersClaimsValidator');
 const helper = require("node-red-viseo-helper");
 const botmgr = require("node-red-viseo-bot-manager");
 
-
 let adapter;
-let bot;
+let skillEndpoint;
+let skillHandler;
 
-async function initConnector(config, node, startCmd) {
-  return new Promise( function (resolve, reject) {
+const getAuthConfig = (config, allowedSkills) => {
+  if (config.botType === 'rootBot' || config.botType === 'skillBot') {
+    const allowedCallersClaimsValidator = new AllowedCallersClaimsValidator(allowedSkills);
+
+    return new AuthenticationConfiguration([], (claims) => {
+      allowedCallersClaimsValidator.validateClaims(claims);
+    });
+  }
+};
+
+function createAdapter(config, authConfig) {
+  let adapter;
+  let botbuilderConfig;
 
-    let botbuilderConfig = {
+  if (config.botType === 'none' || config.botType === '') {
+    botbuilderConfig = {
+      appId: config.appId,
+      appPassword: config.appPassword,
+    };
+  } else {
+    botbuilderConfig = {
       appId: config.appId,
-      appPassword: config.appPassword
+      appPassword: config.appPassword,
+      authConfig
     };
+  }
+  adapter = new BotFrameworkAdapter(botbuilderConfig);
+
+  return adapter;
+}
+
+function createBot(adapter, config, node, allowedCallers, ifRootBot, authConfig) {
+  let bot;
+
+  // create a simple user-facing bot, neither root nor skill bot
+  if (config.botType === 'none' || config.botType === '') {
+    bot = new VBMBot(node, config.appId, config.startCmd);
```

Comment:
In VBMBot, the 'sendWelcomeMessage' function seems to be always required when you provide the 'welcomeMessage' parameter (here config.startCmd)

Reply:
Don't really understand.

Reply:
the 4th argument of VBMBot constructor is the "sendWelcomeMessage" function. Here, this function is given only if the type is rootBot (line 63). When the type is not given (line 51) or for a skill (line 71), this value is not provided.
In node-red-contrib-botbuilder/msbot/botVBM.js, this function seems to be always required when the 3th argument is provided (welcomeMessage) (line 27). In our case, the 3th argument is always provided (config.startCmd), so it should be the same for the sendWelcomeMessage function.